File: 2020/go/day4.go
package advent2020

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func Day4(inputFile string) error {
	input, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return err
	}

	fmt.Println("Day 4: Passport Processing")
	part1, err := day4Part1(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 1: %d\n", part1)

	part2, err := day4Part2(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 2: %d\n", part2)

	return nil
}

func inputToPassports(input string) []map[string]string {
	passports := []map[string]string{}
	records := strings.Split(input, "\n\n")
	for _, record := range records {
		passport := make(map[string]string)
		record = strings.ReplaceAll(record, "\n", " ")
		fields := strings.Split(record, " ")
		for _, field := range fields {
			kv := strings.Split(field, ":")
			passport[kv[0]] = kv[1]
		}
		passports = append(passports, passport)
	}
	return passports
}

func day4Part1(input string) (int, error) {
	passports := inputToPassports(input)
	valid := 0
	for _, pp := range passports {
		if _, ok := pp["byr"]; !ok {
			continue
		}
		if _, ok := pp["iyr"]; !ok {
			continue
		}
		if _, ok := pp["eyr"]; !ok {
			continue
		}
		if _, ok := pp["hgt"]; !ok {
			continue
		}
		if _, ok := pp["hcl"]; !ok {
			continue
		}
		if _, ok := pp["ecl"]; !ok {
			continue
		}
		if _, ok := pp["pid"]; !ok {
			continue
		}
		valid++
	}
	return valid, nil
}

func validatePassport(passport map[string]string) bool {
	if byr, ok := passport["byr"]; ok {
		if len(byr) != 4 {
			return false
		}
		byri, err := strconv.Atoi(byr)
		if err != nil {
			return false
		}
		if byri < 1920 || byri > 2002 {
			return false
		}
	} else {
		return false
	}
	if iyr, ok := passport["iyr"]; ok {
		if len(iyr) != 4 {
			return false
		}
		iyri, err := strconv.Atoi(iyr)
		if err != nil {
			return false
		}
		if iyri < 2010 || iyri > 2020 {
			return false
		}
	} else {
		return false
	}
	if eyr, ok := passport["eyr"]; ok {
		if len(eyr) != 4 {
			return false
		}
		eyri, err := strconv.Atoi(eyr)
		if err != nil {
			return false
		}
		if eyri < 2020 || eyri > 2030 {
			return false
		}
	} else {
		return false
	}
	if hgt, ok := passport["hgt"]; ok {
		u := hgt[len(hgt)-2:]
		vv := hgt[:len(hgt)-2]
		v, err := strconv.Atoi(vv)
		if err != nil {
			return false
		}
		switch u {
		case "cm":
			if v < 150 || v > 193 {
				return false
			}
		case "in":
			if v < 59 || v > 76 {
				return false
			}
		default:
			return false
		}
	} else {
		return false
	}
	if hcl, ok := passport["hcl"]; ok {
		if hcl[0] != '#' {
			return false
		}
		_, err := strconv.ParseInt(hcl[1:], 16, 32)
		if err != nil {
			return false
		}
	} else {
		return false
	}
	if ecl, ok := passport["ecl"]; ok {
		colors := map[string]struct{}{
			"amb": {},
			"blu": {},
			"brn": {},
			"gry": {},
			"grn": {},
			"hzl": {},
			"oth": {},
		}
		if _, ok := colors[ecl]; !ok {
			return false
		}
	} else {
		return false
	}
	if pid, ok := passport["pid"]; ok {
		if len(pid) != 9 {
			return false
		}
		_, err := strconv.Atoi(pid)
		if err != nil {
			return false
		}
	} else {
		return false
	}
	return true
}

func day4Part2(input string) (int, error) {
	passports := inputToPassports(input)
	valid := 0
	for _, passport := range passports {
		if validatePassport(passport) {
			valid++
		}
	}

	return valid, nil
}

File: 2020/rust/src/day4.rs
use simple_error::SimpleError;
use std::collections::HashMap;
use std::collections::HashSet;
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;

pub fn day4(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 4: Passport Processing");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    println!("\tPart 1: {}", part1(&input)?);
    println!("\tPart 2: {}", part2(&input)?);
    Ok(())
}

fn input_to_passports(input: &str) -> Result<Vec<HashMap<String, String>>, Box<dyn Error>> {
    input
        .split("\n\n")
        .try_fold(Vec::new(), |mut passports, record| {
            let passport = record.replace("\n", " ").split(" ").try_fold(
                HashMap::new(),
                |mut passport, field| -> Result<HashMap<String, String>, Box<dyn Error>> {
                    let kv: Vec<&str> = field.split(":").collect();
                    if kv.len() < 2 {
                        return Err(SimpleError::new(format!("Invalid input: {}", record)).into());
                    }
                    passport.insert(kv[0].to_string(), kv[1].to_string());
                    Ok(passport)
                },
            )?;
            passports.push(passport);
            Ok(passports)
        })
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    let fields: HashSet<&str> = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
        .iter()
        .cloned()
        .collect();
    Ok(input_to_passports(input)?
        .iter()
        .fold(0, |valid, passport| {
            let pfields: HashSet<&str> = passport.keys().map(|k| k.as_str()).collect();
            if pfields.intersection(&fields).count() == fields.len() {
                valid + 1
            } else {
                valid
            }
        }))
}

fn passport_valid(passport: &HashMap<String, String>) -> Result<(), Box<dyn Error>> {
    match passport.get("byr") {
        None => return Err(SimpleError::new("").into()),
        Some(byr) => {
            if byr.len() != 4 {
                return Err(SimpleError::new("").into());
            }
            let byr = byr.parse::<i32>()?;
            if byr < 1920 || byr > 2002 {
                return Err(SimpleError::new("").into());
            }
        }
    };
    match passport.get("iyr") {
        None => return Err(SimpleError::new("").into()),
        Some(iyr) => {
            if iyr.len() != 4 {
                return Err(SimpleError::new("").into());
            }
            let iyr = iyr.parse::<i32>()?;
            if iyr < 2010 || iyr > 2020 {
                return Err(SimpleError::new("").into());
            }
        }
    };
    match passport.get("eyr") {
        None => return Err(SimpleError::new("").into()),
        Some(eyr) => {
            if eyr.len() != 4 {
                return Err(SimpleError::new("").into());
            }
            let eyr = eyr.parse::<i32>()?;
            if eyr < 2020 || eyr > 2030 {
                return Err(SimpleError::new("").into());
            }
        }
    };
    match passport.get("hgt") {
        None => return Err(SimpleError::new("").into()),
        Some(hgt) => {
            let unit = &hgt[hgt.len() - 2..];
            let value = hgt[0..hgt.len() - 2].parse::<i32>()?;
            match unit {
                "cm" => {
                    if value < 150 || value > 193 {
                        return Err(SimpleError::new("").into());
                    }
                }
                "in" => {
                    if value < 59 || value > 76 {
                        return Err(SimpleError::new("").into());
                    }
                }
                _ => return Err(SimpleError::new("").into()),
            }
        }
    };
    match passport.get("hcl") {
        None => return Err(SimpleError::new("").into()),
        Some(hcl) => {
            if hcl[0..1].ne("#") {
                return Err(SimpleError::new("").into());
            }
            i32::from_str_radix(&hcl[1..], 16)?;
        }
    }
    match passport.get("ecl") {
        None => return Err(SimpleError::new("").into()),
        Some(ecl) => match &ecl[..] {
            "amb" | "blu" | "brn" | "gry" | "grn" | "hzl" | "oth" => {}
            _ => return Err(SimpleError::new("").into()),
        },
    }
    match passport.get("pid") {
        None => return Err(SimpleError::new("").into()),
        Some(pid) => {
            if pid.len() != 9 {
                return Err(SimpleError::new("").into());
            }
            pid.parse::<i32>()?;
        }
    }
    Ok(())
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    Ok(input_to_passports(input)?
        .iter()
        .fold(0, |valid, passport| match passport_valid(passport) {
            Ok(_) => valid + 1,
            Err(_) => valid,
        }))
}
